prediction_market_agent/agents/think_thoroughly_agent/think_thoroughly_agent.py

```diff
@@ -179,27 +195,38 @@ def generate_final_decision(
         )
         return output
 
-    def answer_binary_market(self, question: str) -> Answer | None:
+    def answer_binary_market(
+        self, question: str, n_iterations: int = 2
+    ) -> Answer | None:
         hypothetical_scenarios = self.get_hypohetical_scenarios(question)
         conditional_scenarios = self.get_required_conditions(question)
 
-        sub_predictions = par_generator(
-            hypothetical_scenarios.scenarios + conditional_scenarios.scenarios,
-            lambda x: (x, self.generate_prediction_for_one_outcome(x)),
-        )
-
-        outcomes_with_probs = []
-        for scenario, prediction in sub_predictions:
-            if prediction is None:
-                continue
-            outcomes_with_probs.append((scenario, prediction))
+        scenarios_with_probs: list[tuple[str, Answer]] = []
+        for iteration in range(n_iterations):
             logger.info(
-                f"'{scenario}' has prediction {prediction.p_yes * 100:.2f}% chance of being True, because: '{prediction.reasoning}'"
+                f"Starting to generate predictions for each scenario, iteration {iteration + 1} / {n_iterations}."
             )
 
+            sub_predictions = par_generator(
+                hypothetical_scenarios.scenarios + conditional_scenarios.scenarios,
+                lambda x: (
+                    x,
+                    self.generate_prediction_for_one_outcome(x, scenarios_with_probs),
```

Comment:
Curious what do you think about this. The idea is to feed into the prompt previous-iteration predictions to "normalise" them against each other.

Reply:
I need more context into the changes to make up my mind.
Do I understand correctly that now you do 2 runs
-> 1st run: do as before, generate sub-predictions with probabilities that don't add to 1
-> 2nd run: pass in all sub-predictions to each agent, so that it can take those non-adding-to-1 probabilities in consideration

Do I understand the above correctly?
If so, I think a better approach would be to have an additional agent (e.g. `NormalizerAgent`), that takes a list of `sub_predictions` and simply normalizes them. This is because, in the current scenario, each agent is tasked with "normalizing" the probability by itself, so I think having only one agent accomplishing this is more consistent.

Curious about what you think.

Reply:
> that takes a list of sub_predictions and simply normalizes them

I was thinking about this too, but let's say the agent receives the following predictions:

- Will GNO hit $1000? p_yes=70%
- Will GNO hit $2000? p_yes=80%

What should the normalisation agent do? Increase the first one to >80% or decrease the second one to <70%? 

Both of the options are just guesses without additional information. So that's why I thought that just providing a list of previous predictions and making the research again should be better.

However, something like a normalisation agent that also does the research could be part of https://github.com/gnosis/prediction-market-agent/issues/108.

Reply:
Yes, happy with adding the normalization on #108 .
Approving this.